TAF specification (#373)
openlawlibrary/taf: specification/taf-spec.md

  ├─ snapshot.json (v2)
  └─ timestamp.json (v2)
- targets/
  │
  └─ repository.json
```

Finally, let's assume that a new signing key had to be added, which means that `root.json` had
to be modified and signed:

```markdown
Commit SHA: 5678efgh

- metadata/
  │
  ├─ root.json (v2)
  ├─ targets.json (v2)
  ├─ snapshot.json (v3)
  └─ timestamp.json (v3)
- targets/
  │
  └─ repository.json
```

To ensure ease of use in managing and maintaining authentication repositories, TAF is equipped with a
series of Command Line Interface (CLI) commands. These commands are designed to automate the setup
process, making it accessible and efficient for users. By simply inputting parameters for signing
keys, signature thresholds, and roles, the CLI commands facilitate the automatic creation of
authentication repositories. Additionally, for the purpose of updating repositories, TAF's CLI
commands are tailored to simplify the modification of target and metadata files. This includes tasks
such as updating the expiration dates of metadata files. The commands also enable the addition of new
roles, the integration of new target and reference authentication repositories within the TAF

> framework.

Reviewer: Do we also want to add hyperlink to the[ command line specification](https://github.com/openlawlibrary/taf/blob/master/docs/repo-creation-and-update-util.md)?

Reviewer: If we aren't sure about including the command-line specification link here, then I'd vote to remove.

## 4 Tracking valid states of target repositories

This section explains how TAF tracks changes across multiple target repositories.
TAF treats all target repositories as parts of a single, integrated system. The authentication
repository is key in this process, as it records the combined state of all target repositories at
any given time. By examining the commit history of the authentication repository, we can
view and understand the evolving state of this system of repositories over time.

As previously mentioned, target repositories are specified in `repositories.json`, and through
the combined information stored in this file along with `mirrors.json`, their URLs are defined.

> The updater uses this information in order to initially clone these repositories. The specifics of
> this cloning process will be discussed in greater detail in a later section.

Reviewer: ```suggestion
The updater uses this information in order to initially clone these repositories.
```

> Now, let's consider an example where we have three target repositories: `namespace1/repo1`,

Reviewer: ```suggestion
Let's consider an example where we have three target repositories: `namespace1/repo1`,
```

> `namespace1/repo2`, and `namespace1/repo3` and freshly initialized authentication

Reviewer: ```suggestion
`namespace1/repo2`, `namespace1/repo3`, an initialized authentication
```

> repository containing TUF metadata files and initial target files (for simplicity, let's assume
> that there are only the two necessary files, `repositories.json` and `mirrors.json` and that there
> are no references to other authentication repositories):

Reviewer: ```suggestion
repository containing TUF metadata files and initial target files. Let's assume
that there are the two necessary files, `repositories.json` and `mirrors.json` and that there
are no references to other authentication repositories:
```

`branch1` of `repo3` as well). The current state of these target repositories would now look like
this:

```markdown
namespace1/
│
├── repo1/
│   └── main: a1b2c3d4, e3f4g5h6
│
├── repo2/
│   ├── main: e5f6g7h8
│   └── branch1: i7j8k9l0
│
└── repo3/
    ├── main: i9j0k1l2
    └── branch1: m2n3o4p5
```

The next revision of the authentication repository is then:

```markdown
- metadata/
  ├─ root.json
  ├─ targets.json
  ├─ snapshot.json
  └─ timestamp.json
- targets/
  ├─ repositories.json
  └─ mirrors.json
  └─ namespace1/
      ├─ repo1
      │   └─ {"branch": "main", "commit": "e3f4g5h6"}
      ├─ repo2
      │   └─ {"branch": "branch1", "commit": "e5f6g7h8"}
      └─ repo3
          └─ {"branch": "branch1", "commit": "m2n3o4p5"}
```

By navigating through the commit history of the authentication repository, it's possible to
collect all registered commits for each target repository and each of their branches. So, within
the TAF framework, there's no need to track all past commits of target repositories in a single
revision of the authentication repository. Similarly, previous versions of metadata files can be

> easily accessed through the Git history.

Reviewer: Excellent example!

## 5 Updater

> This section focuses on the Updater, a key component of the TAF framework. The Updater's main job is

Reviewer: We sometimes use Updater with capital U, but most of the time I've seen the updater in small case. Could we standardize?

out-of-band commit. The out-of-band commit, which for now is expected to be first commit of the
authentication repository, is obtained by contacting the official publishers of the repository. If
this commit is provided, it acts as a verification step to ensure there's no mismatch between the
expected initial commit and the actual first commit in the remote repository. This verification
helps to detect scenarios where users might be directed to download a repository from untrustworthy
sources, such as those created by malicious actors.

The update process in TAF can be broken down into the following steps:

1. Clone the authentication repository as a bare Git repository to a temporary location. Validate its
compliance with TUF specification at every revision, starting from the previously saved last
validated commit. Check the out-of-band commit if provided.
1. Update the user's existing local repository copy or create a new one if it doesn't exist.
1. Clone target repositories if they are not present locally, or fetch their changes. Do not merge new changes.
1. Compare the data in the authentication repository with the fetched commits of target repositories.

> 1. Execute any scripts designated to run after the update process

Reviewer: These are all 1.s

After each successful update, TAF stores the last validated commit of the authentication repository
in a file on the user's local machine. This file is placed in a folder named with the authentication
repository's name, prefixed with an underscore (_). The purpose of saving this commit is to optimize
future updates by avoiding starting from the first commit each time.

> Note: At the moment, there are no implemented security measures which would make it harder to
> manually modify the file storing this last validated commit. That should be addressed in the future.
> We have not yet thoroughly thought about what would be the best way to do that.

Reviewer: ```suggestion
Note: At present, there are no implemented security measures that prevent a user from manually modifying the last validated commit locally. We will be focusing on validating more local repository states as part of future security assurance. 
```

#### 5.1.1 Validate the authentication repository

> The update process begins with cloning the authentication repository to a temporary location as a
> bare Git repository. If a URL is provided when calling the updater, the repository is cloned from
> that location. If no URL is provided but a user's local repository already exists, the URL is
> determined based on its remote. In a bare Git repository setup:

Reviewer: ```suggestion
The update process begins with cloning the authentication repository to a temporary location as a
bare Git repository. If a remote URL is provided when calling the updater, the repository is cloned from
that remote. If no remote URL is provided, the URL is
determined based on Git repository remote from the user's local repository. In a bare Git repository setup:
```

authentication repository is up-to-date with the latest validated state and that this state
is accurately recorded for future reference.

An important aspect of TAF's update process is the determination of destination paths for target
repositories on the filesystem. When initiating the updater, an optional input parameter can be
provided that specifies the root location for all repositories. For instance, if this root location
is set to /dir1/dir2, the path for a target repository will be formed by appending its full name to
this path. This would result in a path like /dir1/dir2/namespace/name1 for the target repository.

In cases where this root path is not explicitly provided, TAF defaults to determining it based on the
path of the authentication repository. It assumes that the full name of the authentication repository
is included in its path, and the root for the target repositories is located two directories up from
the authentication repository's location."

Suppose the path of the authentication repository is `/dir1/dir2/namespace/auth_repo`.
Considering that the authentication repository's full name (`namespace/auth_repo`) is included

> in its path, TAF calculates the root location for target repositories as `/dir1/dir2`.

Reviewer: These kind of look like implementation details for finding paths on local disk. Are they necessary in the spec? Or should they maybe go into a more detailed document that lists the CLI behavior? Maybe [here ](https://github.com/openlawlibrary/taf/blob/master/docs/repo-creation-and-update-util.md) or [here](https://github.com/openlawlibrary/taf/blob/master/docs/setting-up-new-library.md)?

#### 5.1.3 Execute handlers

TAF incorporates the functionality to execute custom scripts after the updater has run,

> providing flexibility to respond to various outcomes of the update process. To be more precise,

Reviewer: ```suggestion
providing flexibility to respond to various outcomes of the update process. 
```

single authentication repository. In contrast, the `update` folder contains scripts that are
executed once the entire update process for all repositories is complete.

In situations where a TAF authentication repository references other authentication
repositories, scripts in the repo folder will be executed multiple times. This execution occurs
for each update of the referenced authentication repositories, as well as for the top-level
authentication repository itself. On the other hand, scripts in the update folder are triggered
only once, at the very end of the entire update process.

In TAF, each script is expected to output JSON containing two types of data: persistent and
transient. The persistent data is automatically saved to a file named `persistent.json` after
every script execution. This file is then passed on to subsequent scripts. On the other hand,
transient data, while also passed to the next script, is not stored permanently.
In addition to handling persistent and transient data, these scripts also receive detailed
information about the repositories involved — this includes both the authentication repository
and its target repositories.

> 

Reviewer: To avoid repetition here and in updater docs, would it make sense to move this part to the full updater docs? And then we can simplify this section by just saying: for detailed explanation on the scripts, please refer to the Updater documentation and lifecycle hooks [here](https://github.com/openlawlibrary/taf/blob/a35dda05f14d817b6be680bbfabe228793b7fd8a/docs/updater/update_process.md?plain=1#L36)

> ### 5.1.4 Updating repositories with cross-references

Reviewer: We haven't used 'cross-references' in the document, so we should consider using another familiar term

Within TAF, an authentication repository has the capability to reference other authentication
repositories. This relationship is defined in the `dependencies.json` file. Alongside the name of
each referenced repository, the file can also specify out-of-band commits. The URLs for these
referenced authentication repositories are determined similarly to target repositories. This involves
combining their full names in a `namespace/name`format with URL templates specified in `mirrors.
json`.

Once the primary authentication repository (including its own contents and targets) has been
validated without errors, the updater processes each repository listed in the `dependencies.json`. It
determines their URLs and then recursively updates each referenced authentication repository. When an
out-of-band commit is specified for a referenced repository, it is included in the update function.
The updater then conducts a validation check based on this commit, as described earlier.

> 

Reviewer: Now that I've read the 5. Updater section, it does feel like we can move everything that we've written here to the [Updater ](https://github.com/openlawlibrary/taf/blob/master/docs/updater/update_process.md?plain=1#L36) docs, and then reference here in the spec. That would simplify the specification, but also leave it up to the reader who's interested in more details to go over? What are your thoughts?

Author: I think it's fine to reference other documents if there are extensions of the specification, but it should contain everything that describes TAF main functionalities. I don't think that descriptions of how to call the CLI command should go into the specification, but it should be possible to understand how it work. I think that we should figure out if there is anything else that is in these other docs that we should mention here. The old updater specification is very outdated and not needed now that we have this, so I will remove it. The other docs should be focused on how to run these commands, which I think is the case for the most part. Maybe we can reference the specification from other docs instead? Although I do not think that it's necessary to update other docs as a part of this PR

TAF utilizes the four fundamental roles defined by the TUF specification: `root`, `targets`,
`snapshot`, and `timestamp`. Additionally, TAF supports the use of delegated targets roles.
Similarly, the metadata file format in TAF is consistent with the standard established by TUF.
Detailed descriptions can be found in TUF specification:

* [Roles and PKI](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#roles-and-pki--roles-and-pki)
* [Metadata Format](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#metaformat--metaformat)

In TAF, target files preserve their original definition from TUF. On the other hand,
TUF allows flexibility in choosing filenames and directory structures for target files within a
repository, while TAF sets clear guidelines. It requires a specific directory structure, mandates target
files named after their respective target repositories, and requires additional target files that link
an authentication repository with its target repositories and other authentication repositories.
Furthermore, TAF supports the use of YubiKeys, although it does not make their use mandatory.

> The following are high-level actors their corresponding actions within the framework:

Reviewer: ```suggestion
The following are high-level actors and their corresponding actions within the framework:
```

That might be possible after transitioning to new TUF. However, it's crucial to always ensure backward compatibility. In addition
to the planned transition to a new version of TUF, the potential implementation of TAP 19 could impact the need for
separate target files that store commit SHAs and branches. For more details, see [Section 2.1](#21-content-addressable-systems-and-tuf)."

### [3.2 Target files](#32-target-files)

TAF requires certain types of target files in order to perform validation of repositories.
Modifying these target files involves updating and signing TUF metadata files. This process ensures a high
level of security, as only individuals possessing the signing keys are authorized to make changes.
It is worth noting that TAF allows existence of additional target files, not included in this
overview, if a specific use-case requires them.

Note: if information from this section is moved to metadata files, the section will be updated. However,
maintaining backward compatibility will always be necessary. In addition to transitioning to a new
version of TUF, as mentioned earlier, if TAP 19 is implemented, there might no longer be a need

> for separate target files storing commit SHAs and branches. See [seciton 2.1](#21-content-addressable-systems-and-tuf)

Reviewer: ```suggestion
for separate target files storing commit SHAs and branches. See [section 2.1](#21-content-addressable-systems-and-tuf)
```

out-of-band commit. The out-of-band commit, which for now is expected to be first commit of the
authentication repository, is obtained by contacting the official publishers of the repository. If
this commit is provided, it acts as a verification step to ensure there's no mismatch between the
expected initial commit and the actual first commit in the remote repository. This verification
helps to detect scenarios where users might be directed to download a repository from untrustworthy
sources, such as those created by malicious actors.

The update process in TAF can be broken down into the following steps:

1. Clone the authentication repository as a bare Git repository to a temporary location. Validate its
compliance with TUF specification at every commit, starting from the previously saved last
validated commit. Check the out-of-band commit if provided.
1. Update the user's existing local repository copy or create a new one if it doesn't exist.
1. Clone target repositories if they are not present locally, or fetch their changes. Do not merge new changes.
1. Compare the data in the authentication repository with the fetched commits of target repositories.

> 1. Execute any scripts designated to run after the update process

Reviewer: These are still 1.s

Author: I didn't see that comment. This is actually correct, if you view the rendered file, it will have the correct enumeration.
